In [1]:
import numpy as np

# task1

In [2]:
class TASK_1:
    
    def SearchRotatedSortedArray(self,A,t):
        l = 0
        r = len(A) -1
        while l <= r:
            mid = (l+r)//2
            if A[mid] == t: 
                return mid
            if A[mid] < A[r]:
                if A[mid] < t and A[r] >= t:
                    l = mid +1 
                else: 
                    r = mid - 1
            else:
                if A[l] <= t and A[mid] > t:
                    r = mid -1
                else: 
                    l = mid + 1
        return -1

    def getKth(self, A, B, k):
        k = int(k)
        lenA = len(A);
        lenB = len(B)
        if lenA > lenB: return self.getKth(B, A, k)
        if lenA == 0: return B[k - 1]
        if k == 1: return min(A[0], B[0])
        pa = int(min(k / 2, lenA))
        pb = k - pa
        if A[pa - 1] <= B[pb - 1]:
            return self.getKth(A[pa:], B, pb)
        else:
            return self.getKth(A, B[pb:], pa)

    def findMedianSortedArrays(self, A, B):
        lenA = len(A);
        lenB = len(B)
        if (lenA + lenB) % 2 == 1:
            return self.getKth(A, B, (lenA + lenB) / 2 + 1)
        else:
            return (self.getKth(A, B, (lenA + lenB) / 2) + self.getKth(A, B, (lenA + lenB) / 2 + 1)) * 0.5

    def findMedianSortedArrays2(self, A, B):
        lenA = len(A);
        lenB = len(B)
        return (self.getKth(A, B, (lenA + lenB + 1) / 2) + self.getKth(A, B, (lenA + lenB + 2) / 2)) * 0.5

    
    def two_sum(self,A,t):
        rel = []
        for i in range(len(A) - 1):
            for j in range(i + 1, len(A)):
                if A[i] + A[j] == t:
                    rel.append(i)
                    rel.append(j)
                    return rel


    def Palindrome_Number(self,n):
        Palindrome = str(n)
        if Palindrome == Palindrome[::-1]:
            return "this is a Palindrome number!"
        else:
            return "this is not a Palindrome number!"
        
if __name__ == '__main__':
    t = 9
    a = [1,3,5,7]
    b = [2,4,6,8,9,10]
    c = [2,7,11,15]
    PN = 96569
    
    print("num1:",a)
    print("num2:",b)
    print("The index of target is:",TASK_1().SearchRotatedSortedArray(b,t))
    print("#####################")
    print("The Median of Two Sorted Arrays is:",TASK_1().getKth(b,a,5))
    print("#####################")
    print("The sum of index in nums is:",TASK_1().two_sum(c,t))
    print("#####################")
    print("The Palindrome Number is:",PN,". So", TASK_1().Palindrome_Number(t))    

num1: [1, 3, 5, 7]
num2: [2, 4, 6, 8, 9, 10]
The index of target is: 4
#####################
The Median of Two Sorted Arrays is: 5
#####################
The sum of index in nums is: [0, 1]
#####################
The Palindrome Number is: 96569 . So this is a Palindrome number!


# TAKS2

In [3]:
from numpy import *

def Find_solution_linsys(a,b):
    at = np.linalg.inv(a)
    if np.linalg.matrix_rank(a) < len(a[:,0]):
        print ("the linear algebraic equations has not solution! ")
    else:
        Solution = np.array(at.dot(b))
        return(Solution)

a = np.array([[8, -6, 2],[-4, 11, -7],[4, -7, 6]])
b = np.array([[28],[-40],[33]])
Find_solution_linsys(a,b)

array([[ 2.],
       [-1.],
       [ 3.]])

# task3

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

In [5]:
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

We can see that there are many values of the character type in the dataset, so we need to perform data processing on the data.

In [6]:
print(pd.isnull(train).sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


# Some description and analysis of the data set

For the sex feature:
We can use 0 to male and 1 to female

In [7]:
sex_mapping = {"male": 0, "female": 1}
train['Sex'] = train['Sex'].map(sex_mapping)
test['Sex'] = test['Sex'].map(sex_mapping)

In feature of Age, is missing 177 values, but the Age is important to survival, so we need to fill these values.Not only that, We should construct the training data by segmenting the ages and counting the number of people in each age group. I had not thought of a suitable way to deal with the age profile, so I found such a way to deal with the age profile.
Here is the link https://www.kaggle.com/startupsci/titanic-data-science-solutions

In [8]:
#sort the ages into logical categories
train["Age"] = train["Age"].fillna(-0.5)
test["Age"] = test["Age"].fillna(-0.5)
bins = [-1, 0, 5, 12, 18, 24, 35, 60, np.inf]
labels = ['Unknown', 'Baby', 'Child', 'Teenager', 'Student', 'Young Adult', 'Adult', 'Senior']
train['AgeGroup'] = pd.cut(train["Age"], bins, labels = labels)
test['AgeGroup'] = pd.cut(test["Age"], bins, labels = labels)

In [9]:
#create a combined group of both datasets
combine = [train, test]

#extract a title for each Name in the train and test datasets
for dataset in combine:
    dataset['Title'] = dataset.Name.str.extract(' ([A-Za-z]+)\.', expand=False)

pd.crosstab(train['Title'], train['Sex'])

Sex         0    1
Title             
Capt        1    0
Col         2    0
Countess    0    1
Don         1    0
Dr          6    1
Jonkheer    1    0
Lady        0    1
Major       2    0
Master     40    0
Miss        0  182
Mlle        0    2
Mme         0    1
Mr        517    0
Mrs         0  125
Ms          0    1
Rev         6    0
Sir         1    0

In [10]:
#replace various titles with more common names
for dataset in combine:
    dataset['Title'] = dataset['Title'].replace(['Lady', 'Capt', 'Col',
    'Don', 'Dr', 'Major', 'Rev', 'Jonkheer', 'Dona'], 'Rare')
    
    dataset['Title'] = dataset['Title'].replace(['Countess', 'Lady', 'Sir'], 'Royal')
    dataset['Title'] = dataset['Title'].replace('Mlle', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Ms', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Mme', 'Mrs')

train[['Title', 'Survived']].groupby(['Title'], as_index=False).mean()

Title  Survived
0  Master  0.575000
1    Miss  0.702703
2      Mr  0.156673
3     Mrs  0.793651
4    Rare  0.285714
5   Royal  1.000000

In [11]:
#map each of the title groups to a numerical value
title_mapping = {"Mr": 1, "Miss": 2, "Mrs": 3, "Master": 4, "Royal": 5, "Rare": 6}
for dataset in combine:
    dataset['Title'] = dataset['Title'].map(title_mapping)
    dataset['Title'] = dataset['Title'].fillna(0)

train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1  38.0      1      0   
2                             Heikkinen, Miss. Laina    1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1  35.0      1      0   
4                           Allen, Mr. William Henry    0  35.0      0      0   

             Ticket     Fare Cabin Embarked     AgeGroup  Title  
0         A/5 21171   7.2500   NaN        S      Student      1  
1          PC 17599  71.2833   C85        C        Adult      3  
2  STON/O2. 3101282   7.9250   NaN        S  Young Adult      2  
3            113803  53.1000  C123        S  Young Adult      3  
4            373450   8.0500   NaN        S  Young Adult      1

In [12]:
# fill missing age with mode age group for each title
mr_age = train[train["Title"] == 1]["AgeGroup"].mode() #Young Adult
miss_age = train[train["Title"] == 2]["AgeGroup"].mode() #Student
mrs_age = train[train["Title"] == 3]["AgeGroup"].mode() #Adult
master_age = train[train["Title"] == 4]["AgeGroup"].mode() #Baby
royal_age = train[train["Title"] == 5]["AgeGroup"].mode() #Adult
rare_age = train[train["Title"] == 6]["AgeGroup"].mode() #Adult

age_title_mapping = {1: "Young Adult", 2: "Student", 3: "Adult", 4: "Baby", 5: "Adult", 6: "Adult"}

#I tried to get this code to work with using .map(), but couldn't.
#I've put down a less elegant, temporary solution for now.
#train = train.fillna({"Age": train["Title"].map(age_title_mapping)})
#test = test.fillna({"Age": test["Title"].map(age_title_mapping)})

for x in range(len(train["AgeGroup"])):
    if train["AgeGroup"][x] == "Unknown":
        train["AgeGroup"][x] = age_title_mapping[train["Title"][x]]
        
for x in range(len(test["AgeGroup"])):
    if test["AgeGroup"][x] == "Unknown":
        test["AgeGroup"][x] = age_title_mapping[test["Title"][x]]

In [13]:
#map each Age value to a numerical value
age_mapping = {'Baby': 1, 'Child': 2, 'Teenager': 3, 'Student': 4, 'Young Adult': 5, 'Adult': 6, 'Senior': 7}
train['AgeGroup'] = train['AgeGroup'].map(age_mapping)
test['AgeGroup'] = test['AgeGroup'].map(age_mapping)

train.head()

#dropping the Age feature for now, might change
train = train.drop(['Age'], axis = 1)
test = test.drop(['Age'], axis = 1)

And for Embarked , we need to fill values someway. There are only 3 value in our feature Embarked, We can fill it with the one with the most of these three values.

In [14]:
print("Number S :")
southampton = train[train["Embarked"] == "S"].shape[0]
print(southampton)

print("Number C:")
cherbourg = train[train["Embarked"] == "C"].shape[0]
print(cherbourg)

print("Number Q:")
queenstown = train[train["Embarked"] == "Q"].shape[0]
print(queenstown)

Number S :
644
Number C:
168
Number Q:
77


In [15]:
train = train.fillna({"Embarked": "S"})
embarked_mapping = {"S": 1, "C": 2, "Q": 3}
train['Embarked'] = train['Embarked'].map(embarked_mapping)
test['Embarked'] = test['Embarked'].map(embarked_mapping)

We can drop some useless feature 
And for the feature Cabin , is missing 687 values, i think there shouldnt have a big relationships between house number and survived, so i think we can drop this feature.
For the Ticket and Name feature, we can also drop it. 

In [16]:
train = train.drop(['Name'], axis = 1)
test = test.drop(['Name'], axis = 1)
train = train.drop(['Cabin'], axis = 1)
test = test.drop(['Cabin'], axis = 1)
train = train.drop(['Ticket'], axis = 1)
test = test.drop(['Ticket'], axis = 1)

Due to the differences in ticketing paths and ticketing methods, the fare feature does not describe some things, and I think the fare feature can be removed directly, as the Pclass feature tells us the class of cabin the passenger is travelling in, which is directly bound by the price. 

In [17]:
train = train.drop(['Fare'], axis = 1)
test = test.drop(['Fare'], axis = 1)

In [18]:
train.head()

PassengerId  Survived  Pclass  Sex  SibSp  Parch  Embarked  AgeGroup  Title
0            1         0       3    0      1      0         1       4.0      1
1            2         1       1    1      1      0         2       6.0      3
2            3         1       3    1      0      0         1       5.0      2
3            4         1       1    1      1      0         1       5.0      3
4            5         0       3    0      0      0         1       5.0      1

In [19]:
test.head()

PassengerId  Pclass  Sex  SibSp  Parch  Embarked  AgeGroup  Title
0          892       3    0      0      0         3       5.0      1
1          893       3    1      1      0         1       6.0      3
2          894       2    0      0      0         3       7.0      1
3          895       3    0      0      0         1       5.0      1
4          896       3    1      1      1         1       4.0      3

# model prediction  

I will use a machine learning model of decision trees to make predictions about survival rates.

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# split data 
predictors = train.drop(['Survived', 'PassengerId'], axis=1)
target = train["Survived"]
x_train, x_val, y_train, y_val = train_test_split(predictors, target, test_size = 0.22, random_state = 0)

#predictrion 
decisiontree = DecisionTreeClassifier()
decisiontree.fit(x_train, y_train)
y_pred = decisiontree.predict(x_val)
x_val.insert(x_val.shape[1],'Pre_Survived',y_pred)
pd.set_option('display.max_rows', None)
print(x_val)

     Pclass  Sex  SibSp  Parch  Embarked  AgeGroup  Title  Pre_Survived
495       3    0      0      0         2       5.0      1             0
648       3    0      0      0         1       5.0      1             0
278       3    0      4      1         3       2.0      4             0
31        1    1      1      0         2       6.0      3             1
255       3    1      0      2         2       5.0      3             1
298       1    0      0      0         1       5.0      1             0
609       1    1      0      0         1       6.0      2             1
318       1    1      0      2         1       5.0      2             0
484       1    0      1      0         2       5.0      1             1
367       3    1      0      0         2       6.0      3             1
704       3    0      1      0         1       5.0      1             0
346       2    1      0      0         1       6.0      2             1
196       3    0      0      0         3       5.0      1       

# Evaluation:

In [21]:
#I used this metric to evaluate the result of prediction. 
acc = round(accuracy_score(y_pred, y_val) * 100, 2)
print(acc)

78.68
